aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -238,6 +238,36 @@ SparseTensor& resize_as_sparse_(SparseTensor& self, const SparseTensor& src) {
   return self;
 }
 
+SparseTensor dense_to_sparse(const Tensor& self){
+  return dense_to_sparse(self, self.dim());
+}
+
+SparseTensor dense_to_sparse(const Tensor& self, int64_t sparseDims){
+  int64_t dims = self.dim();
+  AT_CHECK(sparseDims > 0, "sparseDims must be >0");
+  AT_CHECK(sparseDims <= dims, 
+    "sparseDims must be less than or equal to self.dim()");
+  at::TensorOptions spO = self.type().toSparse().options();
+  std::vector<int64_t> sizes = self.sizes().vec();
+
+  Tensor nz = self.nonzero().transpose(0, 1);
+  LongTensor indices;
+  if(sparseDims == dims){
+    indices = nz.clone();
+  }else{
+    Tensor i = nz.narrow(0, 0, sparseDims);
+    std::tie(indices, std::ignore) = _unique_dim(i, 1);
+    indices = indices.clone();
```

Comment:
indices returned from `_unique_dim()` should be a copy already, why do you need an extra `clone` here?

Reply:
I do not know why an extra clone is needed actually.  Not doing so creates some unexpected (to me) behavior which I documented here: https://github.com/pytorch/pytorch/issues/12633

Reply:
I see, the reason is indices returned from `_unique_dim()` is not contiguous. In that case, maybe you can `indices = indices.contiguous()`

Reply:
Is this something that should be fixed in ```_unique_dim```?

Reply:
IIUC, `_unique_dim` doesn't have to assure returned tensor to be contiguous. So I don't think `_unique_dim` needs to be fixed. Functions that take in tensors should not assume contiguous, if needed, just call `contiguous` or modify kernel to make it work with non-contiguous cases.